### DISCUSSION TITLE
Change version number strategy

### DISCUSSION THREAD

**Discussion by bethesque** [pact-foundation/pact-broker-docker#102] [Ideas]
Currently the version number strategy uses the M.m.p from the pact_broker gem, and adds an extra number to the end to signify the release. eg. 2.106.1.1 is for gem version 2.106.1 and release 1 (releases start at 0). The release number is bumped when the pact_broker gem hasn't changed, but something related to the Docker image packaging has (eg. ruby version bump, alpine version bump, other gem dependency bump, changing or fixing something to do with the application setup etc)

When the base docker image (alpine) is upgraded, there may be breaking changes introduced (eg. at a certain alpine version, support for Docker < 20 was dropped).

The current versioning scheme does not allow us to indicate the presence of breaking changes to the underlying Docker image without the pact_broker gem itself (which as no knowledge of the Docker project) incrementing a major version.

The current tagging system does have the advantage, however, of allowing easy identification of the pact_broker version from the tag - it would be good to maintain that.

Proposal - start a new versioning scheme using that allows breaking changes to be conveyed for either the pact_broker gem OR some aspect of the underlying Docker image, while also including the pact_broker gem version.

Proposed format: `M.m-pactbroker<pact_broker_gem_version>`

The `M.m` would be incremented if there were changes to either the packaging or the gem, as appropriate.

The current major version is 2, so to allow the numbers to keep incrementing, I suggest we start with `3.0-pactbroker2.106.0` (or whatever gem version we're at at the time of the change).

**Comment by TimothyJones** [comment]
The [semver spec](https://semver.org/) already covers this, I think.

The proposed format with the dash makes it look like a prerelease version from a semver perspective, which I think is potentially confusing. I would say that the underlying broker version would more correctly be listed as build metadata (since it's information about what is built, and not a pre-release indicator).

For this reason, I think `<Major>.<Minor>.<Patch>+pactbroker.v<pact_broker_gem_version>` would be the most appropriate format.

Unfortunately, docker tags don't support `+` as a character (which is annoying, since people often use the tags to put semver version numbers in 🤷 ). As a compromise, you could do something like:

```
<Major>.<Minor>.<Patch>_with_pactbroker.v<pact_broker_gem_version>
```  

to make it clearer that it's _not_ a prerelease (or a standard semver tag, generally).

**Reply by TimothyJones** [reply]
For those playing along at home, people have been complaining about docker tags not being semver-compatible [for a number of years now](https://github.com/distribution/distribution/issues/1201). There's an [open issue](https://github.com/opencontainers/distribution-spec/issues/154) against the relevant spec, but it doesn't look like it has much traction :(

**Reply by bethesque** [reply]
Yeah, I looked at making it a proper semantic version, but the lack of support for `+` killed that.

**Reply by bethesque** [reply]
The idea is to follow the docker conventions I've seen in docker images like these (node and ruby):

<img width="784" alt="Screenshot 2023-05-01 at 11 53 24 am" src="https://user-images.githubusercontent.com/446228/235389655-c44ed70b-2a61-4504-889a-5040b2b94dbf.png">
<img width="770" alt="Screenshot 2023-05-01 at 11 53 43 am" src="https://user-images.githubusercontent.com/446228/235389666-fbcaa48e-902e-4b6f-97fa-1d3936b99dfe.png">

**Reply by TimothyJones** [reply]
As far as I can see, that would cover all the cases. I like it!
